monty-31/src/aarch64_neon/poseidon2.rs

```diff
@@ -318,13 +350,78 @@ where
     fn permute_state_terminal(&self, state: &mut [PackedMontyField31Neon<FP>; WIDTH]) {
         external_terminal_permute_state(
             state,
-            self.external_constants.get_terminal_constants(),
-            add_rc_and_sbox_generic,
+            &self.packed_terminal_external_constants,
+            add_rc_and_sbox::<FP, D>,
             &MDSMat4,
         );
     }
 }
 
+/// Converts a scalar constant into a packed NEON vector in "negative form".
+///
+/// This is a pre-computation step for round constants. Instead of storing a constant `c`
+/// and performing a modular addition `(val + c) mod P` in the permutation loop, we pre-compute
+/// `c' = c - P`. This allows the round update to be a simple, non-modular signed integer
+/// addition `val + c'`, which is significantly faster.
+#[inline(always)]
+fn convert_to_vec_neg_form_neon<MP: MontyParameters>(input: i32) -> uint32x4_t {
+    unsafe {
+        // Perform the subtraction as a standard integer operation.
+        //
+        // This computes the "negative form" of the constant.
+        let input_sub_p = input - (MP::PRIME as i32);
```

Comment:
Honestly I think this is self explanatory enough to not need the comment.
```suggestion
        let input_sub_p = input - (MP::PRIME as i32);
```